[vcpkg] Support building vcpkg.exe with mingw

Currently, `bootstrap-vcpkg.bat` requires Visual Studio installed to build `vcpkg.exe`.
It would be nice to support building `vcpkg.exe` with mingw without Visual Studio installed.
Detect `g++` present in `%PATH%`.

I have same error:
```
D:\Download\vcpkg-master\toolsrc>cmake --build build
[  1%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/commands.edit.cpp.obj
D:\Download\vcpkg-master\toolsrc\src\vcpkg\commands.edit.cpp: In function 'void vcpkg::Commands::Edit::perform_and_exit(const vcpkg::VcpkgCmdArguments&, const vcpkg::VcpkgPaths&)':
D:\Download\vcpkg-master\toolsrc\src\vcpkg\commands.edit.cpp:212:62: error: no matching function for call to 'u8path(__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >&, const __gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >&)'
  212 |             candidate_paths.push_back(fs::u8path(first, comma));
      |                                                              ^
In file included from D:/Download/vcpkg-master/toolsrc/include/pch.h:5,
                 from <command-line>:
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:46:10: note: candidate: 'std::filesystem::__cxx11::path fs::u8path(vcpkg::StringView)'
   46 |     path u8path(vcpkg::StringView s);
      |          ^~~~~~
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:46:10: note:   candidate expects 1 argument, 2 provided
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:47:17: note: candidate: 'std::filesystem::__cxx11::path fs::u8path(const char*, const char*)'
   47 |     inline path u8path(const char* first, const char* last) { return u8path(vcpkg::StringView{first, last}); }
      |                 ^~~~~~
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:47:36: note:   no known conversion for argument 1 from '__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >' to 'const char*'
   47 |     inline path u8path(const char* first, const char* last) { return u8path(vcpkg::StringView{first, last}); }
      |                        ~~~~~~~~~~~~^~~~~
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:48:17: note: candidate: 'std::filesystem::__cxx11::path fs::u8path(std::initializer_list<char>)'
   48 |     inline path u8path(std::initializer_list<char> il) { return u8path(vcpkg::StringView{il.begin(), il.end()}); }
      |                 ^~~~~~
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:48:17: note:   candidate expects 1 argument, 2 provided
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:49:17: note: candidate: 'std::filesystem::__cxx11::path fs::u8path(const char*)'
   49 |     inline path u8path(const char* s) { return u8path(vcpkg::StringView{s, s + ::strlen(s)}); }
      |                 ^~~~~~
D:/Download/vcpkg-master/toolsrc/include/vcpkg/base/files.h:49:17: note:   candidate expects 1 argument, 2 provided
mingw32-make[2]: *** [CMakeFiles\vcpkglib.dir\build.make:545: CMakeFiles/vcpkglib.dir/src/vcpkg/commands.edit.cpp.obj] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:100: CMakeFiles/vcpkglib.dir/all] Error 2
mingw32-make: *** [Makefile:114: all] Error 2
```

--- a/toolsrc/include/vcpkg/base/files.h
+++ b/toolsrc/include/vcpkg/base/files.h
@@ -48,14 +48,10 @@ namespace fs
     inline path u8path(std::initializer_list<char> il) { return u8path(vcpkg::StringView{il.begin(), il.end()}); }
     inline path u8path(const char* s) { return u8path(vcpkg::StringView{s, s + ::strlen(s)}); }
 
-    // this allows us to support implementations where std::is_same_v<std::string::const_iterator, const char*>
-    // otherwise, this would collide with u8path(const char*, const char*)
-    template <class T, class = std::enable_if_t<std::is_convertible_v<T, std::string::const_iterator>>>
-    inline path u8path(T first, T last)
+    inline path u8path(std::string::const_iterator first, std::string::const_iterator last)
     {
-        std::string::const_iterator first_it = first, last_it = last;
-        auto firstp = &*first_it;
-        return u8path(vcpkg::StringView{firstp, firstp + (last_it - first_it)});
+        auto firstp = &*first;
+        return u8path(vcpkg::StringView{firstp, firstp + (last - first)});
     }
 
     std::string u8string(const path& p);

--- a/toolsrc/include/vcpkg/base/files.h
+++ b/toolsrc/include/vcpkg/base/files.h
@@ -48,20 +48,15 @@ namespace fs
     inline path u8path(std::initializer_list<char> il) { return u8path(vcpkg::StringView{il.begin(), il.end()}); }
     inline path u8path(const char* s) { return u8path(vcpkg::StringView{s, s + ::strlen(s)}); }
 
-#if defined(_MSC_VER)
-    inline path u8path(std::string::const_iterator first, std::string::const_iterator last)
+    // this allows us to support implementations where std::is_same_v<std::string::const_iterator, const char*>
+    // otherwise, this would collide with u8path(const char*, const char*)
+    template <class T, class = std::enable_if_t<std::is_convertible_v<T, std::string::const_iterator>>>
+    inline path u8path(T first, T last)
     {
-        if (first == last)
-        {
-            return path{};
-        }
-        else
-        {
-            auto firstp = &*first;
-            return u8path(vcpkg::StringView{firstp, firstp + (last - first)});
-        }
+        std::string::const_iterator first_it = first, last_it = last;
+        auto firstp = &*first_it;
+        return u8path(vcpkg::StringView{firstp, firstp + (last_it - first_it)});
     }
-#endif
 
     std::string u8string(const path& p);
     std::string generic_u8string(const path& p);

--- a/toolsrc/src/vcpkg/base/files.cpp
+++ b/toolsrc/src/vcpkg/base/files.cpp
@@ -104,6 +104,11 @@ namespace
 
 fs::path fs::u8path(vcpkg::StringView s)
 {
+    if (s.size() == 0)
+    {
+        return fs::path();
+    }
+
 #if defined(_WIN32)
     return fs::path(vcpkg::Strings::to_utf16(s));
 #else

